glean-core/ios/Glean/Scheduler/MetricsPingScheduler.swift

```diff
@@ -0,0 +1,216 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/// MetricsPingScheduler facilitates scheduling the periodic assembling of metrics pings,
+/// at a given time, trying its best to handle the following cases:
+///
+/// - ping is overdue (due time already passed) for the current calendar day;
+/// - ping is soon to be sent in the current calendar day;
+/// - ping was already sent, and must be scheduled for the next calendar day.
+class MetricsPingScheduler {
+    // This struct is used for organizational purposes to keep the class constants in a single place
+    struct Constants {
+        static let logTag = "glean/MetricsPingSched"
+        static let lastMetricsPingSentDateTime = "last_metrics_ping_iso_datetime"
+        static let dueHourOfTheDay = 4
+        static let lastVersionOfAppUsed = "last_version_of_app_used"
+    }
+
+    private let logger = Logger(tag: Constants.logTag)
+
+    var timer: Timer?
+
+    /// Schedules the metrics ping collection at the due time.
+    ///
+    /// - parameters:
+    ///     * now:  A `Date` representing the current date/time
+    ///     * sendTheNextCalendarDay:  Determines whether to schedule collection for the next calendar day
+    ///                            or to attempt to schedule it for the current calendar day. If the latter and
+    ///                            we're overdue for the expected collection time, the task is scheduled for
+    ///                            immediate execution.
+    func schedulePingCollection(_ now: Date, sendTheNextCalendarDay: Bool) {
+        var fireDate = Calendar.current.date(
+            bySettingHour: Constants.dueHourOfTheDay,
```

Comment:
Weird API.  Was confused by `bySetting` on hour, but not the others, but seems to do the same thing for each of them...

Reply:
Swift (and its predecessor ObjC) have this thing where they combine the first argument "hour" with _what_ it should do, and following arguments don't have that.

It does read weird if you don't know it, but once you do it kinda reads like a sentence (I am _not_ the biggest fan of this convention)

Reply:
Swift often strives to read like a sentence, so you get some interesting parameter/function names sometimes.  In this case, you use the `Calendar.current` to generate a date with some adjusted elements.